File: native/probe/hively/probe_hively.c
/* Does HivelyTracker's replayer play the AHX and HVL files we removed from the app?
 *
 * `SupportedFormats.kt` claimed `ahx` and `hvl` until 2026-09-04, when they were removed because
 * nothing in the app could play them — libopenmpt does not, and neither does anything else we
 * vendor. That is roughly 1,433 Modland files the app used to list and now does not, and it is the
 * one gap in the format table that is our own doing rather than a decision about scope.
 *
 * HivelyTracker's `hvl2wav/` is a standalone replayer: `replay.c`, `replay.h`, `types.h`, BSD
 * 3-Clause. This probe asks the questions `Backend` will have to answer, not "does it make noise":
 *
 *   1. **Does it load from a buffer?** The app never has a path — SAF hands it bytes. `hvl_reset`
 *      is the buffer entry point and dispatches AHX ("THX") and HVL itself, so the header gate the
 *      backend needs is already there and is used here rather than restated.
 *   2. **How many subsongs?** `ht_SubsongNr`. The app's subsong model is the one part of the
 *      player that has been wrong twice, so the count is measured, not assumed.
 *   3. **Does the tune ever end?** `ht_SongEndReached`. Without it a file plays until the user
 *      stops it, which is what `gme` did before it got a fade.
 *   4. **Is it audible?** Peak sample over the render, because "loads" is not "plays" — that was
 *      the whole lesson of the gme probe, where track 0 of a KSS is often an empty slot.
 *
 * One further thing this build is looking for. `types.h` typedefs `uint32` to `unsigned long`,
 * which is 32 bits on the Amiga this came from and **64 bits on every target we ship**. If any of
 * the replayer's arithmetic depends on wrapping at 32 bits, it breaks there and not here — so the
 * probe is built both ways (see `build-hively-probe.sh`) and the verdicts compared.
 *
 * Prints one verdict per file:
 *   VERDICT full subsongs=N ends=yes|no len=NNNs peak=NNNNN version=N title="..."
 *   VERDICT silent subsongs=N ...  |  VERDICT reject:<reason>  |  VERDICT unsupported
 *
 * `len` is how far it got before `ht_SongEndReached`, or the whole window if it never did -- which
 * is the number that decides whether the app can show a duration or has to invent one.
 *
 * Build: ./scripts/build-hively-probe.sh
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "types.h"
#include "replay.h"

/* The engine's rate. The replayer emits one PAL frame per call, so the buffer size is its choice,
 * not ours -- which is itself a finding: the backend will need a ring buffer, unlike the other five. */
enum { RATE = 44100, DEFAULT_SECONDS = 8 };

static int seconds = DEFAULT_SECONDS;

static int8 frame[(RATE * 2 * 2) / 50];

/** Renders one subsong for a moment and reports its loudest sample, or -1 if it will not start. */
static long peakOfSubsong(struct hvl_tune *ht, int subsong, int limit, int *ended, long *played) {
    if (!hvl_InitSubsong(ht, (uint32) subsong)) return -1;
    ht->ht_SongEndReached = 0;

    const long frames = (long) limit * 50;
    long peak = 0;
    long f = 0;
    for (; f < frames; f++) {
        if (ht->ht_SongEndReached) { if (ended) *ended = 1; break; }
        hvl_DecodeFrame(ht, frame, &frame[2], 4);
        const short *samples = (const short *) frame;
        for (size_t i = 0; i < sizeof(frame) / sizeof(short); i++) {
            const long v = samples[i] < 0 ? -(long) samples[i] : samples[i];
            if (v > peak) peak = v;
        }
    }
    if (played) *played = f;                      /* PAL frames, so 50 to the second */
    return peak;
}

/** The title with quotes and control characters removed, so one verdict stays one line. */
static void printTitle(const char *name) {
    for (const unsigned char *p = (const unsigned char *) name; *p && p - (const unsigned char *) name < 64; p++)
        putchar(*p < 0x20 || *p == '"' ? ' ' : *p);
}

int main(int argc, char **argv) {
    int allSubsongs = 0;
    while (argc > 1 && argv[1][0] == '-') {
        if (strcmp(argv[1], "--subsongs") == 0) { allSubsongs = 1; argv++; argc--; }
        else if (strcmp(argv[1], "--seconds") == 0 && argc > 2) { seconds = atoi(argv[2]); argv += 2; argc -= 2; }
        else { fprintf(stderr, "usage: probe-hively [--subsongs] [--seconds N] FILE\n"); return 2; }
    }
    if (argc < 2) { fprintf(stderr, "usage: probe-hively [--subsongs] [--seconds N] FILE\n"); return 2; }

    FILE *file = fopen(argv[1], "rb");
    if (!file) { puts("VERDICT unreadable"); return 2; }
    fseek(file, 0, SEEK_END);
    const long length = ftell(file);
    fseek(file, 0, SEEK_SET);
    uint8 *bytes = length > 0 ? malloc((size_t) length) : NULL;
    if (!bytes || fread(bytes, 1, (size_t) length, file) != (size_t) length) {
        fclose(file); free(bytes); puts("VERDICT unreadable"); return 2;
    }
    fclose(file);

    /* The same gate `hvl_reset` applies, checked here so a rejection is a verdict and not a
     * "Invalid file." on stdout from inside the library. */
    if (length < 16 ||
        !((bytes[0] == 'T' && bytes[1] == 'H' && bytes[2] == 'X' && bytes[3] < 3) ||
          (bytes[0] == 'H' && bytes[1] == 'V' && bytes[2] == 'L' && bytes[3] < 2))) {
        free(bytes);
        puts("VERDICT unsupported");
        return 1;
    }

    hvl_InitReplayer();

    /* freeit = 0: the buffer stays ours, which is the ownership the backend wants. */
    struct hvl_tune *ht = hvl_reset(bytes, (uint32) length, 4, RATE, 0);
    if (!ht) { free(bytes); puts("VERDICT reject:load"); return 1; }

    const int subsongs = (int) ht->ht_SubsongNr + 1;   /* ht_SubsongNr counts the extras */
    int ended = 0;
    long played = 0;
    const long peak = peakOfSubsong(ht, 0, seconds, &ended, &played);

    if (peak < 0) {
        printf("VERDICT reject:subsong0 subsongs=%d\n", subsongs);
    } else {
        printf("VERDICT %s subsongs=%d ends=%s len=%lds peak=%ld version=%d title=\"",
               peak > 0 ? "full" : "silent", subsongs, ended ? "yes" : "no",
               played / 50, peak, (int) ht->ht_Version);
        printTitle(ht->ht_Name);
        puts("\"");
    }

    if (allSubsongs) {
        for (int s = 0; s < subsongs; s++) {
            int subEnded = 0;
            long subPlayed = 0;
            const long p = peakOfSubsong(ht, s, 2, &subEnded, &subPlayed);
            printf("  subsong %d peak=%ld ends=%s\n", s, p, subEnded ? "yes" : "no");
        }
    }

    hvl_FreeTune(ht);
    free(bytes);
    return 0;
}

File: scripts/probe-hively.py
#!/usr/bin/env python3
"""Measure whether HivelyTracker's replayer plays the AHX and HVL files the app dropped.

`ahx` and `hvl` were in `SupportedFormats.kt` until 2026-09-04 and were removed because nothing we
vendor plays them — the list was claiming files the app would then fail to open. That is the only
gap in the format table that is our own doing, so it is the one worth closing, and closing it starts
the way the other five did: a number on this machine before a line of app code.

Two things are measured, not one:

  **does it play**   loads from a buffer, first subsong audible, and whether the tune ever ends
                     -- and if it does, how long it is, which is a duration the app would otherwise
                     have to invent. The render is roughly 2000x realtime, so the window is ten
                     minutes rather than the eight seconds this started with; at eight, almost
                     nothing had ended yet and the answer looked like "it never ends".
  **is it portable** the same files through a second build whose `uint32` is 32 bits rather than
                     the host's 64. Upstream's `types.h` says `unsigned long`, which was 32 bits on
                     the Amiga this replayer comes from. If the two builds ever disagree, the
                     replayer depends on 32-bit wrapping and the app must carry the narrow
                     typedefs; if they never do, that is one less thing to find out on a phone.

Modland names these both ways round — `AHX.title` and `title.ahx` — which is why the survey counts
prefixes as well as extensions. Getting that wrong is what made UADE's reach look like 807 files
when it was 29,127.

Usage:  ./scripts/probe-hively.py [--files N] [--seed N]
"""
from __future__ import annotations

import argparse
import collections
import importlib.util
import json
import pathlib
import random
import subprocess
import sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
CACHE = pathlib.Path.home() / ".protracktor" / "uade-probe"
BUILD = ROOT / "native" / "probe" / "hively" / "build"
BUILDS = {"native": BUILD / "probe-hively-native", "stdint": BUILD / "probe-hively-stdint"}

# What the app would claim again. Both are the replayer's own two header magics, THX and HVL.
NAMES = ["ahx", "hvl"]


def shared():
    spec = importlib.util.spec_from_file_location("probe_uade", ROOT / "scripts" / "probe-uade.py")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def marker(path: str) -> str | None:
    """Which of the names this file carries, at either end. Modland uses both conventions."""
    name = path.rsplit("/", 1)[-1].lower()
    if "." not in name:
        return None
    head, tail = name.split(".", 1)[0], name.rsplit(".", 1)[-1]
    if tail in NAMES:
        return tail
    if head in NAMES:
        return head
    return None


def ask(binary: pathlib.Path, path: pathlib.Path, seconds: int) -> str:
    try:
        result = subprocess.run([str(binary), "--seconds", str(seconds), str(path)],
                                capture_output=True, text=True, timeout=300)
        line = next((l[len("VERDICT "):].strip() for l in result.stdout.splitlines()
                     if l.startswith("VERDICT ")), "")
        return line or (f"exit{result.returncode}" if result.returncode else "crash")
    except subprocess.TimeoutExpired:
        return "timeout"
    except Exception as error:                                       # noqa: BLE001
        return f"error:{error}"


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--files", type=int, default=30)
    parser.add_argument("--seed", type=int, default=68)
    parser.add_argument("--seconds", type=int, default=600,
                        help="how long to render before giving up on a song end (hvl2wav uses 600)")
    args = parser.parse_args()

    missing = [name for name, path in BUILDS.items() if not path.is_file()]
    if missing:
        print(f"❌ not built: {', '.join(missing)}. Run ./scripts/build-hively-probe.sh first.")
        return 1

    s = shared()
    entries = s.modland_index()
    by_name: dict[str, list[str]] = collections.defaultdict(list)
    conventions: collections.Counter = collections.Counter()
    for _size, path in entries:
        found = marker(path)
        if found:
            by_name[found].append(path)
            leaf = path.rsplit("/", 1)[-1].lower()
            conventions[f"{found}:{'suffix' if leaf.rsplit('.', 1)[-1] == found else 'prefix'}"] += 1

    total = sum(len(v) for v in by_name.values())
    print(f"Modland holds {total:,} files these two names claim:\n")
    for name in NAMES:
        print(f"  {name:<5} {len(by_name.get(name, [])):>7,}")
    print("\n  by convention: " + ", ".join(f"{k}×{v:,}" for k, v in sorted(conventions.items())))
    print()

    rng = random.Random(args.seed)
    results: dict[str, collections.Counter] = {}
    details: dict[str, list[tuple[str, str, str]]] = {}
    for name in NAMES:
        paths = by_name.get(name) or []
        if not paths:
            continue
        sample = rng.sample(sorted(paths), min(args.files, len(paths)))
        counts: collections.Counter = collections.Counter()
        rows: list[tuple[str, str, str]] = []
        print(f"=== {name} — {len(sample)} of {len(paths):,}")
        for i, path in enumerate(sample, 1):
            got = s.download(path)
            if not got:
                counts["unfetchable"] += 1
                continue
            verdicts = {kind: ask(binary, got, args.seconds) for kind, binary in BUILDS.items()}
            counts[verdicts["native"].split(" ", 1)[0]] += 1
            leaf = path.rsplit("/", 1)[-1]
            rows.append((leaf, verdicts["native"], verdicts["stdint"]))
            flag = "" if verdicts["native"] == verdicts["stdint"] else "   ⚠ builds disagree"
            print(f"  [{i}/{len(sample)}] {verdicts['native'].split(' ', 1)[0]:<10} {leaf[:46]}{flag}")
        results[name] = counts
        details[name] = rows

    print("\n=== Result\n")
    grand: collections.Counter = collections.Counter()
    for name, counts in results.items():
        grand.update(counts)
        n = sum(counts.values())
        print(f"  {name:<5} {counts.get('full', 0):>3}/{n:<3}  "
              + ", ".join(f"{k}×{v}" for k, v in counts.most_common()))
    n = sum(grand.values())
    print(f"\n  Overall: {grand.get('full', 0)}/{n} loaded from a buffer and were audible.")

    played = [row for rows in details.values() for row in rows if row[1].startswith("full")]
    ended = [row for row in played if "ends=yes" in row[1]]
    multi = sum(1 for row in played if "subsongs=" in row[1]
                and int(row[1].split("subsongs=")[1].split()[0]) > 1)
    print(f"  Of those, {len(ended)}/{len(played)} reached a song end within {args.seconds}s, "
          f"and {multi} have more than one subsong.")
    if ended:
        lengths = sorted(int(row[1].split("len=")[1].split("s")[0]) for row in ended)
        mid = lengths[len(lengths) // 2]
        print(f"  Their lengths run {lengths[0]}s to {lengths[-1]}s, median {mid}s — a duration the"
              "\n  app can show without a song-length database.")

    def shape(verdict: str) -> str:
        """Everything the app depends on, with the peak left out.

        The peak is the loudest sample of a render; two builds whose noise generator wraps at
        different widths will differ there by a few counts and agree about the music. What must not
        differ is the verdict itself, the subsong count, whether it ends and how long it is."""
        return " ".join(field for field in verdict.split() if not field.startswith("peak="))

    hard = [row for rows in details.values() for row in rows if shape(row[1]) != shape(row[2])]
    soft = [row for rows in details.values() for row in rows
            if row[1] != row[2] and shape(row[1]) == shape(row[2])]
    print()
    if hard:
        print(f"  ⚠ {len(hard)} files disagree between the 64-bit and 32-bit builds about something"
              "\n    the app depends on. armeabi-v7a is one of our three ABIs, so this ships:")
        for leaf, native, stdint in hard[:10]:
            print(f"      {leaf}\n        long   {native}\n        int32  {stdint}")
    else:
        print("  The 64-bit and 32-bit builds agree on every file about what plays, how many"
              "\n  subsongs it has, whether it ends and how long it is.")
    if soft:
        print(f"  {len(soft)} of them render a slightly different peak — the two widths do not wrap"
              "\n  the same way — so arm64 and armeabi-v7a will not be sample-identical.")

    (CACHE / "hively.json").write_text(json.dumps(
        {name: dict(counts) for name, counts in results.items()}, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main())
